### ISSUE TITLE
the py-editor is created repeatedly

### ISSUE THREAD

**Issue by sch246** [pyscript/pyscript#1946]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help

### What happened?

When I use VuePress to create some py-editors, it initially runs well on my local machine. 

However, when I deploy it to a server or use `https://pyscript.net/releases/2024.1.1/core.js` as a reference, the delay caused by refreshing the page (Especially the first load) always results in the py-editor being created multiple times. 

This problem seems to occur only when there are multiple py-editors and there is network latency.

I tried using 'target' to specify the id, but it didn't work

It looks like this:

![1705565594162](https://github.com/pyscript/pyscript/assets/58353582/a7b2a891-61be-4ebf-ac7d-66ef1adc6bee)

This is my test file, and it does not always produce duplicates. However, when I deploy files on a server, more than one py-editor always results in duplicates, sometimes even more than three.

```html
<!DOCTYPE html>
<html>
  <head>
      <meta charset="utf-8" />
      <meta name="viewport" content="width=device-width,initial-scale=1" />
      <title>PyScript Test</title>
      <link rel="stylesheet" href="https://pyscript.net/releases/2024.1.1/core.css">
      <script type="module" src="https://pyscript.net/releases/2024.1.1/core.js"></script>
  </head>
  <body>
    <script type="py-editor">
      0
    </script>
    <script type="py-editor">
      1
    </script>
    <script type="py-editor">
      2
    </script>
    <script type="py-editor">
      3
    </script>
    <script type="py-editor">
      4
    </script>
    <script type="py-editor">
      5
    </script>
    <!-- more... -->
  <body>
```

![1705564782922](https://github.com/pyscript/pyscript/assets/58353582/186475ab-29cf-402a-adfa-bacafb389a39)

### What browsers are you seeing the problem on? (if applicable)

Chrome, Microsoft Edge

**Comment by WebReflection** [comment]
This is an extremely hard to follow or fix issue as it looks like your end target deployment environment is doing stuff that would not occur otherwise in other servers ... are you using hot-reload or similar stuff in there? That's a "know-nothing" about custom elements approach and it might lead to shenanigans but at least it's a developer feature, not what's landing on production here.

Any extra details on where, with which config, when and/or how, this happens, might help ... the DOM is synchronous by default and cannot invent itself new nodes out of the box so I am really struggling to understand what's going on.

Thanks in advance for any further detail you could provide ... just the fact it works locally would hint to me this is not PyScript doing weird things, it's rather your current host producing an unexpected layout.

**Comment by sch246** [comment]
I saved the above test HTML as a local file and double-clicked to open it, but it still encountered a problem. I think this environment is already simple enough. 

In fact, after I downloaded the project and edited the source code around line 217 in `pyscript/pyscript.core/src/plugins/py-editor.js`, by adding a check for the `-active` suffix, or simply adding the following after `await init(script, type, interpreter);`:
```js
if (script.parentNode) {
    script.parentNode.removeChild(script)
}
```
the problem was resolved.

**Comment by WebReflection** [comment]
but why is that needed? ... that was my original question ... I can't reproduce this locally and I think the code behind the scene has a guard already for already bootstrapped nodes ... why those nodes get bootstrapped multiple times?

If you save the page you are just propagating the issue to whatever page you had the problem already ... we can be defensive about not bootstrapping already known nodes, but it would be bemeficial for everyone to understand why is that the case at all.

**Comment by sch246** [comment]
I apologize for the inconvenience.

I suspect it might be an issue with the browser cache. I've found that when I open the Network panel in the developer console and disable the browser cache, there's a bug that consistently causes the edit box to be duplicated. I'm not sure if this information is helpful for reproducing the problem.

Additionally, regarding my initial situation, it's possible that the issue arose because I frequently had the Network panel in the developer console open while testing web pages, which I did not do when I first started testing locally.

**Comment by WebReflection** [comment]
If I understand correctly, you have an issue only when devtools and network panel is opened, but not when it's closed ... have you tried disabling all the possible Web Extensions you have and see if you can still reproduce?

From our side, what can't be easily reproduced can't be easily tackled or investigated so that until we find the culprit of the issue it's hard to understand. We have integration tests and smoke tests and more to be sure no duplicated bootstraps ever happen and on multiple projects behind the scene too ... again, any extra help would be useful, thanks!

**Comment by JeffersGlass** [comment]
I did some testing with your above example HTML, and I can replicate this issue. Here's my setup:

- Windows 10 Home 19045.3930
-  Chrome 120.0.6099.225
- HTML above served via VS-Code Live Server

Here's my observations:

- When the test page is first opened, or the first time the page is opened _after clearing the browser cache_, multiple copies of some (not all) of the py-editors are created. Perhaps it is every editor except the first... but I'm not confident if that's the case
![image](https://github.com/pyscript/pyscript/assets/1931111/26bbe380-3826-4166-b2f6-dbbb95b1e623)
- Loading the second, third, etc time the page is refreshed/reloaded once it is in the cache, the problem does not cocur (tried 30 refreshes, no issue. Cleared the cache, issue came back).
  - It doesn't feel like a cache issue so much as a timing issue, but that's conjecture.
  
Here's what I see in the inspector when the issue occurs:

```html
<body>
<script type="py-editor-active">0</script>
<py-editor id="py-editor-3" exec-id="0" root="py-editor-3" style="display: block;">...</py-editor>
<script type="py-editor-active-active">1</script>
<py-editor id="py-editor-9" exec-id="0" root="py-editor-9" style="display: block;">...</py-editor>
<py-editor id="py-editor-0" exec-id="0" root="py-editor-0" style="display: block;"></py-editor>
<script type="py-editor-active-active">2</script>
....
```

[Full HTML in this gist](https://gist.github.com/JeffersGlass/e7a1d637755b93f73fe2d999ba446d5a).

So something weird does seem to be going on. The fact that it appears in your environment, server or local, and in mine, is a strong indicator.

**Comment by WebReflection** [comment]
After fixing the PyTerminal I've realized there might be some race condition within the Mutation Observer logic.

Basically PyTerminal and PyEditor pollute the DOM we observe on bootstrap so that the MutationObserver callback might trigger while previous stuff is still initializing, creating duplicates.

I've fixed in a not so ugly way this issue with the PyTerminal and I believe the PyEditor will need a very similar fix.

I will try to have a better look at the code and fix it properly as I don't usually like "*temporary fixes*" so please bear with me, I am on it!

**Comment by WebReflection** [comment]
The MR fixes in a proper way the issue and it's explained what's going on in the MR description itself: https://github.com/pyscript/pyscript/pull/1953

**Comment by WebReflection** [comment]
P.S. that version with latest is already live in our "*canary/dev channel*" so if you grab PyScript from *npm* you should never see the issue again:

  * [core.js](https://cdn.jsdelivr.net/npm/@pyscript/core/dist/core.js) once CDN catches up
  * [core.css](https://cdn.jsdelivr.net/npm/@pyscript/core/dist/core.css) once CDN catches up

**Comment by WebReflection** [comment]
@sch246 @JeffersGlass would you be so kind to confirm the MR that landed in *npm* actually fixes this issue? I am not 100% that's the case as neither CI nor my machine can reproduce this issue (after that MR) but as you apparently could do so, I'd love to have a triple check that's the case as we're planning to release pretty soon, thank you!

**Comment by sch246** [comment]
A simple way to reproduce this is to manually increase network latency, which can be set in Chrome's developer tools.
![image](https://github.com/pyscript/pyscript/assets/58353582/e1386221-3951-4be5-84e2-55f3d98bde40)
I tried this, and it occurred before the fix, but after the fix, this issue no longer appears, at least not in the test files.

**Comment by WebReflection** [comment]
@sch246 I've tried it all, even throttling the CPU over a slower connection ... I can't reproduce **now** (I should've tried harder before) ... the fix seems to work extremely well but it also slightly changed the timing in bootstrap.

I am still not 100% convinced that was a `type` accessor issue because I can't reproduce synthetically any of this but I am glad it's solved so we can merge and, opefully, release!
